src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1080,15 +1080,29 @@ private HttpResponseHeaders CreateResponseHeaders(bool appCompleted)
                 _keepAlive = false;
             }
 
-            // https://tools.ietf.org/html/rfc7230#section-3.3.1
-            // If any transfer coding other than
-            // chunked is applied to a response payload body, the sender MUST either
-            // apply chunked as the final transfer coding or terminate the message
-            // by closing the connection.
-            if (hasTransferEncoding &&
-                HttpHeaders.GetFinalTransferCoding(responseHeaders.HeaderTransferEncoding) != TransferCoding.Chunked)
+            if (hasTransferEncoding)
             {
-                _keepAlive = false;
+                // https://tools.ietf.org/html/rfc7540#section-8.1.2.2
+                // Such intermediaries SHOULD also remove other connection-specific header fields, such as Keep-Alive,
+                // Proxy-Connection, Transfer-Encoding, and Upgrade, even if they are not nominated by the Connection header field.
+                if (_httpVersion > Http.HttpVersion.Http11)
+                {
```

Comment:
Even though it's probably correct to strip connection-specific header fields for HTTP/3 responses, citing the HTTP/2 spec doesn't confirm that.

It's also weird to cite all the other header fields that should be stripped like "Keep-Alive, Proxy-Connection, Transfer-Encoding, and Upgrade" and the "Connection" header itself and then only strip "Transfer-Encoding". We should either strip all of these now or save it for another PR. I'm leaning towards another PR.

Reply:
The quote from the HTTP/3 spec is almost word for word the same.
https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-4.1.1

Yeah, I wanted to keep this PR focused on Transfer-Encoding issues.